src/tyr/route_serializer_osrm.cc

```diff
@@ -1307,6 +1059,263 @@ names_and_refs(const valhalla::DirectionsLeg::Maneuver& maneuver) {
+    auto intersection = json::map({});
+
+    // Get the node and current edge from the enhanced trip path
+    // NOTE: curr_edge does not exist for the arrive maneuver
+    auto node = etp->GetEnhancedNode(i);
+    auto curr_edge = etp->GetCurrEdge(i);
+    auto prev_edge = etp->GetPrevEdge(i);
+
+    // Add the node location (lon, lat). Use the last shape point for
+    // the arrive step
+    auto loc = json::array({});
+    size_t shape_index = arrive_maneuver ? shape.size() - 1 : curr_edge->begin_shape_index();
+    PointLL ll = shape[shape_index];
+    loc->emplace_back(json::fixed_t{ll.lng(), 6});
+    loc->emplace_back(json::fixed_t{ll.lat(), 6});
+    intersection->emplace("location", loc);
+    intersection->emplace("geometry_index", static_cast<uint64_t>(shape_index));
+
+    // Add index into admin list
+    if (node->has_admin_index()) {
+      intersection->emplace("admin_index", static_cast<uint64_t>(node->admin_index()));
+    }
+
+    if (!arrive_maneuver) {
+      if (curr_edge->has_is_urban()) {
+        intersection->emplace("is_urban", curr_edge->is_urban());
+      }
+    }
+
+    auto toll_collection = json::map({});
+    if (node->type() == TripLeg_Node::kTollBooth) {
+      toll_collection->emplace("type", std::string("toll_booth"));
+    } else if (node->type() == TripLeg_Node::kTollGantry) {
+      toll_collection->emplace("type", std::string("toll_gantry"));
+    }
+    if (!toll_collection->empty())
+      intersection->emplace("toll_collection", toll_collection);
+
+    if (node->cost().transition_cost().seconds() > 0)
+      intersection->emplace("turn_duration",
+                            json::fixed_t{node->cost().transition_cost().seconds(), 3});
+    if (node->cost().transition_cost().cost() > 0)
+      intersection->emplace("turn_weight", json::fixed_t{node->cost().transition_cost().cost(), 3});
+    auto next_node = i + 1 < n ? etp->GetEnhancedNode(i + 1) : nullptr;
+    if (next_node) {
+      auto secs = next_node->cost().elapsed_cost().seconds() - node->cost().elapsed_cost().seconds();
+      auto cost = next_node->cost().elapsed_cost().cost() - node->cost().elapsed_cost().cost();
+      if (secs > 0)
+        intersection->emplace("duration", json::fixed_t{secs, 3});
+      if (cost > 0)
+        intersection->emplace("weight", json::fixed_t{cost, 3});
+    }
+
+    // TODO: add recosted durations to the intersection?
+
+    // Add rest_stop when passing by a rest_area or service_area
+    if (i > 0 && !arrive_maneuver) {
+      auto rest_stop = json::map({});
+      for (uint32_t m = 0; m < node->intersecting_edge_size(); m++) {
+        auto intersecting_edge = node->GetIntersectingEdge(m);
+        bool routeable = intersecting_edge->IsTraversableOutbound(curr_edge->travel_mode());
+
+        if (routeable && intersecting_edge->use() == TripLeg_Use_kRestAreaUse) {
+          rest_stop->emplace("type", std::string("rest_area"));
+          if (intersecting_edge->has_sign()) {
```

Comment:
If the `intersecting_edge` has a `Sign`, we extract the sign's `guide_destinations` text, which is the name of the rest-area.